astropy/io/fits/hdu/compressed.py

```diff
@@ -1595,6 +1595,13 @@ def header(self):
                 image_header['EXTNAME'] == self._default_name):
             del image_header['EXTNAME']
 
+        # Remove the PCOUNT GCOUNT cards if the uncompressed header is
+        # from a primary HDU
+        if 'SIMPLE' in image_header and \
+                ('PCOUNT' in image_header or 'GCOUNT' in image_header):
+            del image_header['PCOUNT']
+            del image_header['GCOUNT']
+
```

Comment:
Can you move that above, same place as for the XTENSION change ? (It's the same idea, just different keywords)

Reply:
It was my first try, but https://github.com/astropy/astropy/blob/0446f168dc6e34996482394f00770b52756b8f9c/astropy/io/fits/hdu/compressed.py#L1560-L1573 writes `PCOUNT` and `GCOUNT` in `image_header` even when the compressed keywords are not there... so I need to remove them after.
I'm not sure if this was done to cover a specific case, I preferred to keep them, then.
I will check again, in case my previous attempt I messed up with something...

Reply:
Actually it wouldn't work since P/GCOUNT are set in between... so it seems simpler to do it as you propose :+1:

Reply:
Oh I didn't refresh the page and I'm seeing your comment now ;)
Yeah those lines are probably still needed (though for PCOUNT/GCOUNT I'm not sure since those keywords would be overwritten based on `.data` when writing a file).

Reply:
Thanks for the merge! :)
Just to dig in a bit more, I tested removing the default set for `PCOUNT` and `GCOUNT` when the keywords are deleted at the beginning (https://github.com/astropy/astropy/commit/c27bd6e7b9a897a3d86dfed5737e39aa2b2937bf) and the test passes: but I'm not 100% sure that we are not missing some cases where the keywords are needed.